### ISSUE TITLE
Don't assume the order of continueUserActivity: and applicationDidBecomeActive

### ISSUE THREAD

**Issue by tzahola** [BranchMetrics/ios-branch-deep-linking#699]
While integrating the Branch SDK to our application at Skyscanner, I've found that it fails to deliver deep link parameters if a link (e.g. `https://bby4.test-app.link/foobar`) was received via AirDrop while the application was in the foreground. 

After some debugging, I've found that the root cause of this issue is that the Branch SDK requires the `continueUserActivity:` call to precede `applicationDidBecomeActive` (both are located in `Branch.m`). This is true in most cases, e.g. when our app is not running or in the background, and we open a Universal Link from another app. However, when our app is in the foreground, and we receive a Universal Link via AirDrop, the application becomes active first (when the AirDrop progress dialog is dismissed), and only then do we receive the Universal Link via `continueUserActivity`. 

Both `applicationDidBecomeActive` and `continueUserActivity:` ends up calling `registerInstallOrOpen:` (in `Branch.m`). As we can see in the following snippet, `registerInstallOrOpen:` tries to cleverly avoid doubly registering the install/open event:
```
// If there isn't already an Open / Install request, add one to the queue
    if (![self.requestQueue containsInstallOrOpen])
    {
        BranchOpenRequest *req = [[clazz alloc] initWithCallback:initSessionCallback];

        [self insertRequestAtFront:req];
    }
    // If there is already one in the queue, make sure it's in the front.
    // Make sure a callback is associated with this request. This callback can
    // be cleared if the app is terminated while an Open/Install is pending.
    else
    {
        BranchOpenRequest *req = [self.requestQueue moveInstallOrOpenToFront:self.networkCount];
        req.callback = initSessionCallback;
    }
```
So if `BranchOpenRequest` was already in progress, it won't enqueue a new one. However, consider the following sequence of events:

1. The application becomes active, so we enqueue a `BranchOpenRequest` without any deep link (`self.preferenceHelper.universalLinkUrl` is `nil`)
2. The application receives the `NSUserActivity` from AirDrop, so we pass that to the Branch SDK via `continueUserActivity:`. 
3. The SDK stores the deep link in `self.preferenceHelper.universalLinkUrl`, and tries to enqueue a second `BranchOpenRequest`, however the above snippet prevents it from actually doing it. 

So in the end, the deep link stored in `self.preferenceHelper.universalLinkUrl` won't be sent to the server, because the `BranchOpenRequest` was already sent by the time `continueUserActivity` is called – the link will be ignored. 

Now consider the "regular" case, when the application is in the background (or not running at all), and we open some Universal Link from another app. In this case, we first receive the `NSUserActivity` first, and therefore the initial `BranchOpenRequest` is enqueued with the deep link included, correctly. Only after this does the application become active. 

Please consider looking into this issue, and implementing install/open tracking in a way that doesn't rely on the order of these two application delegate callbacks, as we can see that their order is not guaranteed in general.

**Comment by E-B-Smith** [comment]
Wow -- Great analysis. I'll look into this and you expect a patch in the next release.

**Comment by E-B-Smith** [comment]
I fixed this in version 0.18.8.

https://github.com/BranchMetrics/ios-branch-deep-linking/releases/tag/0.18.8
